### ISSUE TITLE
Use of the dgCMatrix container

### ISSUE THREAD

**Issue by rcastelo** [rcastelo/GSVA#113]
Ensure that the `dgCMatrix` container for single-cell data can be safely provided as input to the gsva() function.

**Comment by moritzschaefer** [comment]
I tried it with `ssgsea`, where it seems to [have worked previously](https://support.bioconductor.org/p/9137823/#9156951), however I get this Error:

*Error in (function (classes, fdef, mtable) : unable to find an inherited method for function ‘wrapData’ for signature ‘"dgCMatrix", "dgCMatrix"’*

Presumably, because `ssgsea` returns a `dgCMatrix` instead of `matrix` object.  

A potential fix could be to force-convert to matrix, if the input is dgCMatrix in `ssgsea`, e.g.:

```
  R <- t(sparseMatrixStats::colRanks(X, ties.method="average"))
  R <- as.matrix(R)  # NEW
  mode(R) <- "integer"
```

This should be fine, given that the derived matrix is probably smaller (n_samples x n_genesets) than the sparse input matrix

**Comment by axelklenk** [comment]
Hi Moritz, thanks for bringing this to our attention.  As you can see from this issue #113, improved handling of sparse matrices is under active development and the problem you reported *should* not occur in `devel`, both on BioConductor and GitHub. -- I'd of course be grateful if you could test and confirm that.

Background: previous versions of `ssgsea()` coerced their `matrix` result to `dgCMatrix` when the input was a `dgCMatrix`.  However, looking at various gene set scoring results, I found that their non-zero rate is >90% and hence not sparse, and coecion to `dgCMatrix` rather disadvantageous (and therefore discontinued per #142 #143 ).

In the current release version, however, your observation is a bug and we'll provide a fix for it asap.

Don't hesitate to get back to us in case of further questions -- and don't forget to include your `sessionInfo()`. :wink:

**Comment by moritzschaefer** [comment]
I tried (installing the current github devel commit) and it works! Thanks a lot!

**Comment by axelklenk** [comment]
We have backported the fix to your bug report and it is available in GSVA 1.50.1 in the current Bioconductor release 3.18.

In addition, I have tracked down the reason for the sparse-to-dense conversion of `dgCMatrix` objects when inside `SingleCellExperiment` (as well as `SummarizedExperiment`) containers but not when passed directly.  A fix is currently being tested and will soon be available in `devel`.
